### PR TITLE
AVRO-3650: [C++] Fix build on Manjaro

### PR DESCRIPTION
**Pull request by jmccl**
## What is the purpose of the change

Fix the build on distros with a newer boost library or compiler - AVRO-3650.

Upgrades C++ version from 11 to 17 - AVRO-3610

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

(More accurately, it now builds.)

### CODE REVIEW

File: lang/c++/CMakeLists.txt [hunk 1]
```diff
@@ -21,7 +21,7 @@ cmake_minimum_required (VERSION 3.1)
 set (CMAKE_LEGACY_CYGWIN_WIN32 0)
 
 if (NOT DEFINED CMAKE_CXX_STANDARD)
-    set(CMAKE_CXX_STANDARD 11)
+    set(CMAKE_CXX_STANDARD 14)
```
Comment (martin-g) [inline lang/c++/CMakeLists.txt:24]: Let's jump directly to 17  - https://issues.apache.org/jira/browse/AVRO-3610

**Review by martin-g — CHANGES_REQUESTED** [review]
The issue description does not have a link to the JIRA ticket. Please follow the template!
The commit messages do not start with `AVRO-3650: `

File: lang/c++/CMakeLists.txt (hunk 1 above)
Comment (jmccl) [reply lang/c++/CMakeLists.txt]: Sure, I've updated the PR.

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @jmccl !

**Pull request merged**